### ISSUE TITLE
Deviation from scipy and R for high OR or low p-value

### ISSUE THREAD

**Issue by agartland** [brentp/fishers_exact_test#47]
Hi
I'm wondering if you've ever run into this issue that your implementation deviates from R and scipy for extreme ORs with low p-values? Example is below. I did notice that scipy seems to use `int64` and even states in a comment that `int32` is not sufficient (though that is with their implementation and I honestly haven't looked hard enough yet to see how yours differs, [scipy note about int32](https://github.com/scipy/scipy/blob/f990b1d2471748c79bc4260baf8923db0a5248af/scipy/stats/_stats_py.py#L5105C10-L5105C10))

We need these low p-values because we are needing to do a multiplicity adjustment and we have a lot of p-values (hence why we need a fast implementation!) Let us know if you have any ideas.
Thanks,
Andrew

scipy and brentp
```
a = 50 * np.ones(10, dtype=np.uint)
b = 1 * np.ones(10, dtype=np.uint)
c = 257 * np.ones(10, dtype=np.uint)
d = 375 * np.ones(10, dtype=np.uint)

res_brentp = fisher.pvalue_npy(a, b, c, d)[2][0]
res_scipy = stats.fisher_exact([[a[0], b[0]], [c[0], d[0]]])[1]
print(f'brentp: {res_brentp}\nscipy: {res_scipy}')

brentp: 1.368624798014407e-06
scipy: 1.2220470363522046e-17
```

R
```
fisher.test( matrix(c(50, 1, 257, 375), nrow = 2))

        Fisher's Exact Test for Count Data

data:  matrix(c(50, 1, 257, 375), nrow = 2)
p-value < 2.2e-16
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
   12.28923 2901.05129
sample estimates:
odds ratio
   72.6098
```

**Comment by brentp** [comment]
I think this is a limitation of the implementation here. If you need p-values that low then you must use scipy. I thought recent versions of scipy were quite fast.

**Comment by dbolser** [comment]
Triage your runs with this package then run scipy for precise values on a manageable set.

@brentp or @tanghaibao, please close if @agartland confirms this is a suitable work around.

**Comment by agartland** [comment]
That's OK by me. I left this mostly as a warning of the limitation to others. I haven't needed it in a while but I think we started using a numba implementation for speed and accuracy.

**Comment by dbolser** [comment]
Good point @agartland. Do you find numba fast enough / faster than this implmentation?

**Comment by dbolser** [comment]
@agartland — apologies, my earlier suggestion here (triage with this package, then rerun scipy for precision) was based on a wrong reading of the problem. Having actually dug in, this is a plain bug and not a limitation of the approach. Thank you for leaving the report up; it was accurate.

The give-away is that the one-tailed value on your table has been correct all along:

```
pvalue(50, 1, 257, 375).right_tail  = 1.2220470363535e-17
scipy.stats.fisher_exact            = 1.2220470363526e-17
```

Twelve significant figures. So the hypergeometric probabilities were never the issue, and neither was int32 — good instinct to check, but that isn't where it went wrong.

What broke is the two-tailed sum. It selects tables with `p <= cutoff + epsilon`, where `epsilon = 1e-6` is an *absolute* slack. Once the observed table's probability falls below that, `cutoff + epsilon` is effectively just `epsilon`, so the condition admits nearly the whole support. The 1.37e-06 you reported isn't a computed value at all — it's the tolerance itself.

Making the tolerance relative (`p <= cutoff * (1 + epsilon)`, which is what R's `fisher.test` does) brings your table to within 1e-12 relative of scipy. PR #60.

On your numba question — if it's serving you well there's no reason to switch back on my account. But the accuracy reason for having moved should be gone once this lands, if you ever want to reconsider.

**Comment by agartland** [comment]
Thanks @dbolser I appreciate the follow-up and the rigor! Nice to know this is a good alternative. Numba can be quirky and adds a dependency, so not always the best option.
